scripts/perfcontrib.py

```diff
@@ -0,0 +1,115 @@
+from json import loads, dumps
+from urllib.request import urlopen, Request
+from urllib.parse import urlencode
+from urllib.error import HTTPError
+import time
+import sys
+import os
+
+tenantId = "72f988bf-86f1-41af-91ab-2d7cd011db47"
+appId = "c2fe4cd0-be4a-468b-aa4f-078c67dcab6e"
+
+uploadService = "https://perfcontrib.azurewebsites.net"
+uploadEndpoint = f"{uploadService}/api/UploadPerfData"
+authEndpoint = f"{uploadService}/.auth/login/aad"
+authDetailsEndpoint = f"{uploadService}/.auth/me"
+
+aadUrl = f"https://login.microsoftonline.com/{tenantId}"
+
+def get_token() -> str:
+    path = os.path.expanduser("~/.perfcontrib")
+
+    token: str | None = None
+
+    try:
+        if not os.path.exists(path):
+            os.makedirs(path)
+        with open(os.path.expanduser("~/.perfcontrib/token")) as tokenfile:
+            token = tokenfile.readline()
+    except FileNotFoundError:
+        pass
+
+    if token:
+        try:
+            with urlopen(Request(authDetailsEndpoint,
+                    headers = { "X-ZUMO-AUTH": token })) as response:
+                print("Using cached credentials.")
+        except HTTPError as error:
+            token = None
+
+    if not token:
+        token = authenticate()
+    
+    if token:
+        with open(os.path.expanduser("~/.perfcontrib/token"), "w") as tokenfile:
+            tokenfile.write(token)
+    
+    return token
+
+def authenticate() -> str:
+    authBody = {
+        "tenant": tenantId,
+        "client_id": appId,
+        "scope": "User.Read openid profile",
+    }
+    
+    authBodyEncoded = urlencode(authBody).encode()
+
+    with urlopen(Request(f"{aadUrl}/oauth2/v2.0/devicecode", data = authBodyEncoded)) as response:
+        devicecodeResponse = loads(response.read().decode('utf-8'))
+
+    print(devicecodeResponse["message"])
+    
+    authBody2 = {
+        "tenant": tenantId, 
+        "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
+        "client_id": appId,
+        "device_code": devicecodeResponse["device_code"]
+    }
+    
+    authBody2Encoded = urlencode(authBody2).encode()
+    
+    authStatus = "waiting"
+    while (authStatus == "waiting"):
+        # Try to get the access token. if we encounter an error check the reason. 
+        # If the reason is we are waiting then sleep for some time. 
+        # If the reason is the user has declined or we timed out then quit.  
+        try:
+            with urlopen(Request(f"{aadUrl}/oauth2/v2.0/token", data = authBody2Encoded)) as response:
+                tokenResponse = loads(response.read().decode('utf-8'))
+            authStatus = "done"
+        except Exception as ex:
+            print("waiting")
+            time.sleep(10)
```

Comment:
What kinds of failures are we expecting here? Are they just transient failures that we just need to keep retrying until it works? Or are there other failures modes to consider that sleeping and retrying doesn't make sense?

Reply:
The only failure I'm really expecting here is that the user hasn't finished logging in using the device code yet.

The design of the device code flow is to print out the message something like:

> To sign in, use a web browser to open then page https://microsoft.com/devicelogin and enter the code XXXXXYYYYZZ to authenticate

and then wait for the user to complete that flow.

Since this is a script running in the terminal of a technical audience, "press Ctrl-C" is the only other terminating condition that I'm realistically expecting, and that's probably not worth handling explicitly.

Will tighten up the comments.